src/probability/kernel/composition.lean

```diff
@@ -0,0 +1,720 @@
+  ... = κ a set.univ * Cη : mul_comm _ _,
+end
+
+instance is_finite_kernel.comp_prod (κ : kernel α β) [is_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_finite_kernel η] :
+  is_finite_kernel (κ ⊗ₖ η) :=
+⟨⟨is_finite_kernel.bound κ * is_finite_kernel.bound η,
+  ennreal.mul_lt_top (is_finite_kernel.bound_ne_top κ) (is_finite_kernel.bound_ne_top η),
+  λ a, calc (κ ⊗ₖ η) a set.univ
+    ≤ (κ a set.univ) * is_finite_kernel.bound η : comp_prod_apply_univ_le κ η a
+... ≤ is_finite_kernel.bound κ * is_finite_kernel.bound η :
+        mul_le_mul (measure_le_bound κ a set.univ) le_rfl (zero_le _) (zero_le _), ⟩⟩
+
+instance is_s_finite_kernel.comp_prod (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  is_s_finite_kernel (κ ⊗ₖ η) :=
+begin
+  rw comp_prod_eq_sum_comp_prod,
+  exact kernel.is_s_finite_kernel_sum (λ n, kernel.is_s_finite_kernel_sum infer_instance),
+end
+
+end composition_product
+
+section map_comap
+/-! ### map, comap -/
+
+variables {γ : Type*} {mγ : measurable_space γ} {f : β → γ} {g : γ → α}
+
+include mγ
+
+/-- The pushforward of a kernel along a measurable function. -/
+noncomputable
+def map (κ : kernel α β) (f : β → γ) (hf : measurable f) : kernel α γ :=
+{ val := λ a, (κ a).map f,
+  property := (measure.measurable_map _ hf).comp (kernel.measurable κ) }
+
+lemma map_apply (κ : kernel α β) (hf : measurable f) (a : α) :
+  map κ f hf a = (κ a).map f := rfl
+
+lemma map_apply' (κ : kernel α β) (hf : measurable f) (a : α) {s : set γ} (hs : measurable_set s) :
+  map κ f hf a s = κ a (f ⁻¹' s) :=
+by rw [map_apply, measure.map_apply hf hs]
+
+lemma lintegral_map (κ : kernel α β) (hf : measurable f) (a : α)
+  {g' : γ → ℝ≥0∞} (hg : measurable g') :
+  ∫⁻ b, g' b ∂(map κ f hf a) = ∫⁻ a, g' (f a) ∂(κ a) :=
+by rw [map_apply _ hf, lintegral_map hg hf]
+
+lemma sum_map_seq (κ : kernel α β) [is_s_finite_kernel κ] (hf : measurable f) :
+  kernel.sum (λ n, map (seq κ n) f hf) = map κ f hf :=
+begin
+  ext a s hs : 2,
+  rw [kernel.sum_apply, map_apply' κ hf a hs, measure.sum_apply _ hs, ← measure_sum_seq κ,
+    measure.sum_apply _ (hf hs)],
+  simp_rw map_apply' _ hf _ hs,
+end
+
+instance is_markov_kernel.map (κ : kernel α β) [is_markov_kernel κ] (hf : measurable f) :
+  is_markov_kernel (map κ f hf) :=
+ ⟨λ a, ⟨by rw [map_apply' κ hf a measurable_set.univ, set.preimage_univ, measure_univ]⟩⟩
+
+instance is_finite_kernel.map (κ : kernel α β) [is_finite_kernel κ] (hf : measurable f) :
+  is_finite_kernel (map κ f hf) :=
+begin
+  refine ⟨⟨is_finite_kernel.bound κ, is_finite_kernel.bound_lt_top κ, λ a, _⟩⟩,
+  rw map_apply' κ hf a measurable_set.univ,
+  exact measure_le_bound κ a _,
+end
+
+instance is_s_finite_kernel.map (κ : kernel α β) [is_s_finite_kernel κ] (hf : measurable f) :
+  is_s_finite_kernel (map κ f hf) :=
+⟨⟨λ n, map (seq κ n) f hf, infer_instance, (sum_map_seq κ hf).symm⟩⟩
+
+/-- Pullback of a kernel, such that for each set s `comap κ g hg c s = κ (g c) s`. -/
```

Comment:
```suggestion
/-- Pullback of a kernel, such that for each set s `comap κ g hg c s = κ (g c) s`.
We include measurability in the assumptions instead of using junk values
to make sure that typeclass inference can infer that the `comap` of a Markov kernel
is again a Markov kernel. -/
```